Makefile

```diff
@@ -1,40 +1,87 @@
 
-project:
-	swift package generate-xcodeproj --output GRPC.xcodeproj
-	@-ruby fix-project-settings.rb GRPC.xcodeproj || echo "Consider running 'sudo gem install xcodeproj' to automatically set correct indentation settings for the generated project."
+### Xcodeproj and LinuxMain
 
+project: ${XCODEPROJ}
+
+${XCODEPROJ}:
+	${SWIFT_PACKAGE} generate-xcodeproj --output $@
+	@-ruby fix-project-settings.rb GRPC.xcodeproj || \
+		echo "Consider running 'sudo gem install xcodeproj' to automatically set correct indentation settings for the generated project."
+
+# Generates LinuxMain.swift, only on macOS.
+generate-linuxmain:
+	${SWIFT_TEST} --generate-linuxmain
+
+### Protobuf Generation ########################################################
+
+# Generates protobufs and gRPC client and server for the Echo example
+generate-echo: plugins
+	protoc Sources/Examples/Echo/echo.proto \
+		--proto_path=Sources/Examples/Echo \
+		--plugin=${SWIFT_BUILD_PATH}/${SWIFT_BUILD_CONFIGURATION}/protoc-gen-swift \
+		--plugin=${SWIFT_BUILD_PATH}/${SWIFT_BUILD_CONFIGURATION}/protoc-gen-swiftgrpc \
+		--swift_out=Sources/Examples/Echo/Generated \
+		--swiftgrpc_out=Sources/Examples/Echo/Generated
+
+### Testing ####################################################################
+
+# Normal test suite.
 test:
-	swift test
+	${SWIFT_TEST}
+
+# Checks that linuxmain has been updated: requires macOS.
+test-generate-linuxmain: generate-linuxmain
+	@git diff --exit-code Tests/LinuxMain.swift Tests/*/XCTestManifests.swift > /dev/null || \
+		{ echo "Generated tests are out-of-date; run 'swift test --generate-linuxmain' to update them!"; exit 1; }
 
-test-plugin:
-	swift build --product protoc-gen-swiftgrpc
-	protoc Sources/Examples/Echo/echo.proto --proto_path=Sources/Examples/Echo --plugin=.build/debug/protoc-gen-swift --plugin=.build/debug/protoc-gen-swiftgrpc --swiftgrpc_out=/tmp
+# Generates code for the Echo server and client and tests them against 'golden' data.
+test-plugin: plugins
+	protoc Sources/Examples/Echo/echo.proto \
+		--proto_path=Sources/Examples/Echo \
+		--plugin=${SWIFT_BUILD_PATH}/${SWIFT_BUILD_CONFIGURATION}/protoc-gen-swift \
+		--plugin=${SWIFT_BUILD_PATH}/${SWIFT_BUILD_CONFIGURATION}/protoc-gen-swiftgrpc \
+		--swiftgrpc_out=/tmp
 	diff -u /tmp/echo.grpc.swift Sources/Examples/Echo/Generated/echo.grpc.swift
 
-test-generate-linuxmain:
-ifeq ($(UNAME_S), Darwin)
-	swift test --generate-linuxmain
-	@git diff --exit-code */LinuxMain.swift */XCTestManifests.swift > /dev/null || { echo "Generated tests are out-of-date; run 'swift test --generate-linuxmain' to update them!"; exit 1; }
-else
-	echo "test-generate-linuxmain is only available on Darwin"
-endif
+### Misc. ######################################################################
 
 clean:
 	-rm -rf Packages
-	-rm -rf .build build
-	-rm -rf GRPC.xcodeproj
-	-rm -rf Package.pins Package.resolved
-	-rm -rf protoc-gen-swift protoc-gen-swiftgrpc
+	-rm -rf ${SWIFT_BUILD_PATH}
```

Comment:
Did you try running `make clean` to ensure that this is complete?

Reply:
Yes, although I omitted `.swiftpm` from this (i.e. support for Swift PM in Xcode 11 when you're not using a generated `xcodeproj`). The rationale being that it's generated and managed by Xcode and not us.